components/esp_websocket_client/esp_websocket_client.c

```diff
@@ -1126,41 +1178,13 @@ int esp_websocket_client_send_with_opcode(esp_websocket_client_handle_t client,
         ESP_LOGE(TAG, "Failed to setup tx buffer");
         goto unlock_and_return;
     }
-    uint32_t current_opcode = opcode;
-    while (widx < len || current_opcode) {  // allow for sending "current_opcode" only message with len==0
-        if (need_write > client->buffer_size) {
-            need_write = client->buffer_size;
-        } else {
-            current_opcode |= WS_TRANSPORT_OPCODES_FIN;
-        }
-        memcpy(client->tx_buffer, data + widx, need_write);
-        // send with ws specific way and specific opcode
-        wlen = esp_transport_ws_send_raw(client->transport, current_opcode, (char *)client->tx_buffer, need_write,
-                                         (timeout == portMAX_DELAY) ? -1 : timeout * portTICK_PERIOD_MS);
-        if (wlen < 0 || (wlen == 0 && need_write != 0)) {
-            ret = wlen;
-            esp_websocket_free_buf(client, true);
-            esp_tls_error_handle_t error_handle = esp_transport_get_error_handle(client->transport);
-            if (error_handle) {
-                esp_websocket_client_error(client, "esp_transport_write() returned %d, transport_error=%s, tls_error_code=%i, tls_flags=%i, errno=%d",
-                                           ret, esp_err_to_name(error_handle->last_error), error_handle->esp_tls_error_code,
-                                           error_handle->esp_tls_flags, errno);
-            } else {
-                esp_websocket_client_error(client, "esp_transport_write() returned %d, errno=%d", ret, errno);
-            }
-            esp_websocket_client_abort_connection(client, WEBSOCKET_ERROR_TYPE_TCP_TRANSPORT);
-            goto unlock_and_return;
-        }
-        current_opcode = 0;
-        widx += wlen;
-        need_write = len - widx;
-
+    if (esp_websocket_client_send_with_exact_opcode(client, opcode | WS_TRANSPORT_OPCODES_FIN, data, len, timeout) != true) {
```

Comment:
So this is the original problem causing https://github.com/espressif/esp-protocols/issues/460

One nitpick here (sorry, missed that) is that this API shouldn't be called `with_exact_opcode()` if we're changing the opcode based on the fragmentation.
So after the prefix, we should at least mention it in the comments.
I'd suggest though adding one extra boolean parameter `append_fin` to make it cleaner.